### ISSUE TITLE
Atomically move to/delete from $HOME/.local/share

### ISSUE THREAD

**Issue by probonopd** [AppImage/libappimage#28]
Currently libappimage directly extracts e.g., desktop files directly to `$HOME/.local/share` (or deletes them from) on each invocation of

* appimage_type1_register_in_system
* appimage_type2_register_in_system
* appimage_unregister_in_system 

Each change in `/.local/share` can trigger local caches such as Sycoca to be re-generated, which results in a _lot_ of CPU activity.

So we should also apply the same logic that @rszibele used in https://github.com/AppImage/appimaged/pull/50 to all files (desktop files, icon files, mime type files and such) that get copied to `HOME/.local/share` for system integration.

We should only _atomically_ (=everything in one go) move files to or delete files from `$HOME/.local/share` after a certain timer of inactivity has passed.

We should do all editing/rewriting of desktop files outside of `HOME/.local/share` in a separate directory.

This will solve:

* https://github.com/AppImage/appimaged/issues/12
* https://github.com/AppImage/appimaged/issues/19

**Comment by TheAssassin** [comment]
Please don't use the term "atomic", it doesn't quite fit here. What you want to do is to move the files "transactionally", i.e., prepare all the move commands and then execute them at once (possibly aborting/reverting if one operation fails).

**Comment by probonopd** [comment]
"All at once" is what I mean, yes.

**Comment by probonopd** [comment]
The library should have a timer. Calling things like `appimage_type1_register_in_system()` should return immediately but not actually perform the integration, but merely write to `/tmp`. Once the timer expires (no more writes went to the temporary location in the last X seconds), everyhting would be copied in at once.

Makes sense?

**Comment by TheAssassin** [comment]
No, all these implicitnesses are actually _really_ bad design. These kinds of usages are something to be handled on a higher level, as they don't fit every use case (only appimaged could make use of this, but for e.g., AppImageLauncher, we want a prompt result, and the application might even _exit_ before that timer expires...). appimagelauncherd for instance handles this just fine on such a higher level, but could definitely benefit from a batch integrator.

**Comment by rszibele** [comment]
I was thinking about adding this in libappimage as I implemented that functionality in appimaged, but I came to the same conclusion that @TheAssassin mentioned, that having that functionality in `appimage_register_in_system` would be a bad idea.

Calling `appimage_register_in_system` and the unregister equivalent should behave exactly as one _would_ expect it to behave (as it does now). Extra functionality should be built on top of it through a batching interface.

My additions to appimaged could then be re-purposed to use the batching mechanism instead of calling `appimage_register_in_system` for each AppImage. However, the timer is something that should be implemented in the daemons themselves.

**Comment by probonopd** [comment]
How can appimaged handle these things when actually all the work happens in `appimage_type1_register_in_system()` which is in `libappimage`?

**Comment by azubieta** [comment]
@rszibele @probonopd this could be done by setting a temporary HOME while calling `appimage_register_in_system` and then moving all at once. But I do agreed that it would better fit into libappimage.

**Comment by probonopd** [comment]
Yes, in `libappimage` please.
